SIMD: Get rid of attribute-based CPU dispatching

seiko2plus opened the pull request:
Attribute-based CPU dispatching is no longer needed, we are entirely counting on dispatchable sources.


TODO:
- [x] benchmark to check if there's a necessity to auto-vectorize AVX512_SKX

Abandon the idea of providing binary objects for AVX512_SKX due to the massive increase
 in binary size over +400k (striped) with no vast change in performance except
 for reciprocal on some data types, tested against gcc/12.2.1 build.

<details>
<summary>Benchmark for AVX512_SKX VS AVX2</summary>

```Bash

    AVX512_SKX         AVX2                    ratio
     [e0e4acb70c]       [0a56560e]
     <removes_old_cpu_dispatcher~1>       <removes_old_cpu_dispatcher>
+     26.0±0.07μs          125±7μs     4.83  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'L')
+     26.0±0.04μs          122±4μs     4.71  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'Q')
+     25.8±0.01μs        110±0.2μs     4.24  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'q')
+     25.9±0.08μs        108±0.6μs     4.18  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'l')
+     26.3±0.04μs      43.7±0.04μs     1.66  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'I')
+     1.63±0.03μs      2.39±0.02μs     1.47  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'l')
+     1.64±0.05μs         2.39±0μs     1.46  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'Q')
+     1.63±0.05μs      2.37±0.01μs     1.45  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'L')
+     1.64±0.04μs      2.37±0.01μs     1.45  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'q')
+     1.63±0.01μs      2.36±0.01μs     1.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'q')
+     1.63±0.01μs         2.35±0μs     1.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'L')
+     1.64±0.02μs      2.36±0.02μs     1.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'Q')
+     1.64±0.01μs         2.35±0μs     1.43  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'l')
+     8.69±0.07μs      12.5±0.03μs     1.43  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'l')
+     8.75±0.09μs      12.4±0.09μs     1.42  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'L')
+     8.65±0.09μs       12.3±0.1μs     1.42  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'Q')
+     8.63±0.02μs       12.2±0.2μs     1.42  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'q')
+        1.63±0μs      2.28±0.01μs     1.39  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
+        1.63±0μs         2.27±0μs     1.39  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
+        1.63±0μs         2.26±0μs     1.39  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'l')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'Q')
+     1.64±0.01μs         2.27±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'L')
+     1.72±0.06μs      2.37±0.01μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'q')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
+     1.74±0.05μs         2.39±0μs     1.37  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
+     1.75±0.08μs      2.37±0.01μs     1.36  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
+     1.74±0.03μs      2.35±0.01μs     1.35  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
+     1.75±0.02μs      2.35±0.01μs     1.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
+     1.75±0.04μs         2.36±0μs     1.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
+     1.77±0.05μs      2.37±0.01μs     1.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
+     1.77±0.02μs         2.35±0μs     1.33  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
+     1.69±0.01μs      2.20±0.01μs     1.31  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'l')
+     1.69±0.02μs      2.20±0.01μs     1.30  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'L')
+     1.69±0.02μs      2.21±0.01μs     1.30  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'Q')
+     1.70±0.02μs      2.21±0.01μs     1.30  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'q')
+     1.70±0.01μs      2.18±0.01μs     1.28  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'l')
+        1.69±0μs      2.15±0.01μs     1.27  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'L')
+     1.70±0.01μs      2.16±0.01μs     1.27  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'Q')
+     1.71±0.01μs      2.17±0.01μs     1.27  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'q')
+         886±7ns         1.10±0μs     1.24  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'H')
+         895±6ns         1.10±0μs     1.23  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'H')
+         897±7ns         1.10±0μs     1.23  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'h')
+     4.25±0.02μs      5.03±0.01μs     1.18  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'I')
+     1.31±0.02μs         1.54±0μs     1.18  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'I')
+     1.31±0.01μs         1.54±0μs     1.17  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'i')
+     4.27±0.08μs         4.95±0μs     1.16  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'i')
+        996±20ns         1.15±0μs     1.16  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'h')
+        1.21±0μs         1.40±0μs     1.15  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'b')
+     1.33±0.03μs      1.54±0.01μs     1.15  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
+        1.19±0μs         1.37±0μs     1.15  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'B')
+     1.34±0.01μs         1.53±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
+     1.32±0.02μs         1.51±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'I')
+        1.19±0μs         1.36±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'b')
+     1.32±0.01μs         1.50±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'i')
+     1.36±0.01μs      1.54±0.01μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'h')
+        1.20±0μs         1.36±0μs     1.13  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'B')
+        1.43±0μs      1.61±0.01μs     1.13  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'H')
+        1.43±0μs         1.61±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'H')
+        1.38±0μs         1.55±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'h')
+     1.34±0.01μs         1.51±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
+     1.35±0.02μs         1.51±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
+      15.3±0.5μs      17.1±0.03μs     1.11  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'l')
+      15.3±0.4μs      17.1±0.06μs     1.11  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'q')
+        1.05±0μs         1.16±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'i')
+     1.41±0.01μs         1.57±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'b')
+     1.34±0.01μs         1.49±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'I')
+        1.34±0μs      1.48±0.01μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'i')
+     1.05±0.01μs         1.16±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'I')
+     1.05±0.01μs         1.16±0μs     1.10  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
+     1.43±0.02μs      1.58±0.01μs     1.10  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'b')
+      15.0±0.3μs      16.6±0.05μs     1.10  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'Q')
+      15.1±0.2μs       16.5±0.1μs     1.10  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'l')
+      15.0±0.3μs      16.5±0.02μs     1.10  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'l')
+        1.06±0μs         1.16±0μs     1.10  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
+     1.50±0.01μs         1.64±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'B')
+     1.51±0.01μs         1.64±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'B')
+        1.60±0μs      1.74±0.01μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'L')
+      15.1±0.5μs      16.4±0.03μs     1.09  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'q')
+     1.34±0.01μs         1.45±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'I')
+        1.60±0μs         1.74±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'Q')
+        1.60±0μs         1.74±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'q')
+        1.61±0μs         1.74±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'l')
+      7.23±0.3μs      7.84±0.06μs     1.08  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'i')
+      15.1±0.3μs      16.4±0.05μs     1.08  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'L')
+     3.28±0.04μs      3.54±0.01μs     1.08  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'h')
+      15.1±0.2μs      16.2±0.02μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'invert'>, 1, 1, 'q')
+     1.14±0.01μs      1.23±0.01μs     1.07  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'h')
+     1.35±0.01μs      1.44±0.01μs     1.07  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'i')
+      15.1±0.1μs      16.2±0.04μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'q')
+      7.09±0.1μs       7.59±0.1μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'positive'>, 1, 1, 'i')
+     2.60±0.03μs      2.78±0.04μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'h')
+     3.29±0.07μs      3.50±0.07μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'conjugate'>, 1, 1, 'H')
+        1.16±0μs      1.23±0.01μs     1.06  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'H')
+     2.57±0.02μs      2.73±0.01μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'H')
+     3.29±0.07μs      3.49±0.03μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'H')
+        1.18±0μs      1.24±0.01μs     1.06  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'H')
+     6.94±0.04μs       7.32±0.1μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'conjugate'>, 1, 1, 'I')
+     1.16±0.01μs         1.23±0μs     1.05  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'h')
+     3.17±0.03μs      3.33±0.06μs     1.05  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'positive'>, 1, 1, 'h')
+        1.15±0μs      1.21±0.01μs     1.05  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'h')
+     1.15±0.01μs         1.21±0μs     1.05  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'H')
-     1.04±0.02μs          992±3ns     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'bitwise_and'>, 1, 1, 1, 'b')
-     1.00±0.01μs          950±1ns     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'b')
-     2.34±0.03μs      2.22±0.01μs     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'bitwise_xor'>, 1, 1, 1, 'l')
-     2.09±0.05μs         1.98±0μs     0.95  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'B')
-      7.29±0.1μs      6.90±0.01μs     0.95  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'i')
-     2.10±0.06μs         1.98±0μs     0.94  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'b')
-     2.34±0.03μs      2.20±0.01μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'l')
-     2.24±0.02μs      2.10±0.01μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'subtract'>, 1, 1, 1, 'Q')
-         639±2ns        600±0.9ns     0.94  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'multiply'>, 1, 1, 1, 'b')
-         640±7ns        596±0.6ns     0.93  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'multiply'>, 1, 1, 1, 'B')
-     2.27±0.08μs         2.10±0μs     0.93  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'subtract'>, 1, 1, 1, 'l')
```
</details>


------

### X86

<details>
<summary>CPU</summary>

```Bash
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         48 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  16
  On-line CPU(s) list:   0-15
Vendor ID:               AuthenticAMD
  Model name:            AMD Ryzen 7 7700X 8-Core Processor
    CPU family:          25
    Model:               97
    Thread(s) per core:  2
    Core(s) per socket:  8
    Socket(s):           1
    Stepping:            2
    Frequency boost:     disabled
    CPU(s) scaling MHz:  67%
    CPU max MHz:         5572.2651
    CPU min MHz:         3000.0000
    BogoMIPS:            8986.45
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep
                         mtrr pge mca cmov pat pse36 clflush mmx fxsr
                          sse sse2 ht syscall nx mmxext fxsr_opt pdpe
                         1gb rdtscp lm constant_tsc rep_good amd_lbr_
                         v2 nopl nonstop_tsc cpuid extd_apicid aperfm
                         perf rapl pni pclmulqdq monitor ssse3 fma cx
                         16 sse4_1 sse4_2 x2apic movbe popcnt aes xsa
                         ve avx f16c rdrand lahf_lm cmp_legacy svm ex
                         tapic cr8_legacy abm sse4a misalignsse 3dnow
                         prefetch osvw ibs skinit wdt tce topoext per
                         fctr_core perfctr_nb bpext perfctr_llc mwait
                         x cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfm
                         on_v2 ibrs ibpb stibp vmmcall fsgsbase bmi1
                         avx2 smep bmi2 erms invpcid cqm rdt_a avx512
                         f avx512dq rdseed adx smap avx512ifma clflus
                         hopt clwb avx512cd sha_ni avx512bw avx512vl
                         xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_o
                         ccup_llc cqm_mbm_total cqm_mbm_local avx512_
                         bf16 clzero irperf xsaveerptr rdpru wbnoinvd
                          cppc arat npt lbrv svm_lock nrip_save tsc_s
                         cale vmcb_clean flushbyasid decodeassists pa
                         usefilter pfthreshold avic v_vmsave_vmload v
                         gif x2avic v_spec_ctrl avx512vbmi umip pku o
                         spke avx512_vbmi2 gfni vaes vpclmulqdq avx51
                         2_vnni avx512_bitalg avx512_vpopcntdq rdpid
                         overflow_recov succor smca fsrm flush_l1d
Virtualization features:
  Virtualization:        AMD-V
Caches (sum of all):
  L1d:                   256 KiB (8 instances)
  L1i:                   256 KiB (8 instances)
  L2:                    8 MiB (8 instances)
  L3:                    32 MiB (1 instance)
NUMA:
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-15
Vulnerabilities:
  Itlb multihit:         Not affected
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Not affected
  Retbleed:              Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disable
                         d via prctl
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __u
                         ser pointer sanitization
  Spectre v2:            Mitigation; Retpolines, IBPB conditional, IB
                         RS_FW, STIBP always-on, RSB filling, PBRSB-e
                         IBRS Not affected
  Srbds:                 Not affected
  Tsx async abort:       Not affected
```
</details>

<details>
<summary>OS</summary>

```Bash
Linux seiko-pc 6.1.1-arch1-1 #1 SMP PREEMPT_DYNAMIC Wed, 21 Dec 2022 22:27:55 +0000 x86_64 GNU/Linux
Python 3.10.9
gcc (GCC) 12.2.1
```
</details>

#### Benchmark

<details>
 <summary>Before vs After (AVX2)</summary>

```Bash
python runtests.py -n --bench-compare parent/main IntContig  -- --cpu-affinity=6,7
```
```Bash   before           after         ratio
       before           after         ratio
     [86450a0c]       [52bf5007]
     <removes_old_cpu_dispatcher~11>       <removes_old_cpu_dispatcher>
+         559±2ns         609±30ns     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'bitwise_xor'>, 1, 1, 1, 'B')
+         563±2ns         600±10ns     1.07  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'bitwise_xor'>, 1, 1, 1, 'b')
+         561±5ns         597±20ns     1.06  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'bitwise_or'>, 1, 1, 1, 'b')
+         555±3ns         584±20ns     1.05  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'bitwise_and'>, 1, 1, 1, 'b')
-     15.9±0.03μs       15.1±0.5μs     0.95  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'L')
-         111±1μs          106±1μs     0.95  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'l')
-     2.39±0.02μs      2.26±0.09μs     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'multiply'>, 1, 1, 1, 'L')
-        2.39±0μs       2.26±0.1μs     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'multiply'>, 1, 1, 1, 'l')
-     2.41±0.02μs       2.27±0.1μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'multiply'>, 1, 1, 1, 'q')
-     15.9±0.06μs       14.9±0.4μs     0.94  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'Q')
-     1.69±0.01μs         1.59±0μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'i')
-     2.42±0.02μs       2.27±0.1μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'multiply'>, 1, 1, 1, 'Q')
-     1.07±0.01μs      1.00±0.01μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'b')
-     2.03±0.03μs      1.79±0.02μs     0.88  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'b')
-     7.74±0.05μs      6.82±0.02μs     0.88  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'I')
-     3.80±0.05μs       3.32±0.1μs     0.87  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'h')
-     21.4±0.04μs       16.2±0.2μs     0.75  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'q')
-     21.5±0.06μs      16.1±0.09μs     0.75  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'l')
-     9.19±0.05μs      6.83±0.07μs     0.74  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'i')
-     3.37±0.02μs      2.16±0.06μs     0.64  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'Q')
-        3.42±0μs      2.17±0.06μs     0.63  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'q')
-     3.51±0.02μs      2.16±0.06μs     0.61  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'L')
-        3.63±0μs      2.22±0.07μs     0.61  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'l')
-     12.8±0.08μs      7.76±0.03μs     0.61  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'i')
-     4.00±0.03μs      2.38±0.02μs     0.59  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'q')
-     3.88±0.01μs      2.30±0.06μs     0.59  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'l')
-        3.13±0μs      1.82±0.04μs     0.58  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'b')
-     4.47±0.01μs      2.57±0.05μs     0.57  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'Q')
-     4.37±0.02μs      2.51±0.03μs     0.57  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'L')
-        5.94±0μs       3.35±0.1μs     0.56  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'h')
-      35.4±0.8μs       16.6±0.3μs     0.47  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'l')
-      35.4±0.7μs       16.5±0.4μs     0.47  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'q')
-     4.03±0.01μs      1.81±0.01μs     0.45  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'I')
-      4.91±0.1μs      2.18±0.06μs     0.44  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'q')
-     3.74±0.01μs         1.66±0μs     0.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'B')
-     6.00±0.02μs      2.64±0.02μs     0.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'q')
-     4.02±0.01μs      1.72±0.02μs     0.43  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'i')
-      34.9±0.4μs       14.9±0.3μs     0.43  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'Q')
-     5.90±0.03μs      2.51±0.03μs     0.43  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'l')
-        5.14±0μs      2.16±0.06μs     0.42  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'Q')
-     3.76±0.01μs         1.56±0μs     0.42  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'b')
-      35.5±0.4μs       14.7±0.4μs     0.41  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'L')
-        5.34±0μs      2.16±0.06μs     0.40  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'l')
-        5.38±0μs      2.16±0.06μs     0.40  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'L')
-        3.41±0μs      1.37±0.01μs     0.40  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'B')
-     6.29±0.03μs      2.52±0.01μs     0.40  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'q')
-     6.42±0.02μs      2.57±0.04μs     0.40  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'Q')
-     6.31±0.01μs      2.52±0.03μs     0.40  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'L')
-     4.11±0.02μs         1.63±0μs     0.40  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'H')
-     4.33±0.01μs      1.70±0.02μs     0.39  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'i')
-     6.18±0.01μs      2.41±0.04μs     0.39  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'l')
-     4.14±0.01μs         1.58±0μs     0.38  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'b')
-        3.37±0μs      1.27±0.02μs     0.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'i')
-        3.43±0μs      1.27±0.02μs     0.37  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'I')
-     6.38±0.01μs      2.28±0.01μs     0.36  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
-        6.41±0μs         2.28±0μs     0.36  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
-     6.40±0.01μs         2.28±0μs     0.36  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
-        6.41±0μs         2.27±0μs     0.35  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
-        3.98±0μs      1.36±0.01μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'b')
-        6.96±0μs      2.37±0.01μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
-     6.99±0.01μs      2.37±0.02μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
-        6.98±0μs      2.36±0.01μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
-     7.00±0.01μs      2.37±0.02μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
-     6.98±0.01μs      2.36±0.01μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
-        6.98±0μs      2.36±0.03μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
-     6.99±0.02μs      2.36±0.02μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
-        6.98±0μs      2.35±0.02μs     0.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
-     5.29±0.02μs      1.64±0.02μs     0.31  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'H')
-        3.68±0μs         1.12±0μs     0.31  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'H')
-        3.71±0μs         1.13±0μs     0.30  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'H')
-     5.96±0.02μs      1.79±0.01μs     0.30  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'I')
-     5.71±0.01μs         1.66±0μs     0.29  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'B')
-        5.53±0μs         1.56±0μs     0.28  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'h')
-     5.68±0.01μs         1.57±0μs     0.28  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'h')
-        5.16±0μs         1.37±0μs     0.27  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'B')
-     5.38±0.01μs         1.40±0μs     0.26  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'b')
-        5.14±0μs      1.26±0.02μs     0.25  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'I')
-        5.22±0μs      1.27±0.02μs     0.24  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'i')
-        4.94±0μs         1.12±0μs     0.23  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'h')
-     6.96±0.01μs      1.53±0.01μs     0.22  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
-     6.96±0.01μs      1.52±0.01μs     0.22  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
-     6.94±0.01μs      1.52±0.02μs     0.22  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
-     6.94±0.01μs      1.51±0.03μs     0.22  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
-     5.38±0.02μs         1.16±0μs     0.22  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'h')
-        6.37±0μs      1.18±0.01μs     0.19  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
-     6.36±0.01μs      1.17±0.01μs     0.18  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
-        6.91±0μs         1.22±0μs     0.18  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'h')
-        6.91±0μs      1.22±0.01μs     0.18  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'H')
-        6.91±0μs      1.22±0.02μs     0.18  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'H')
-     6.90±0.02μs      1.20±0.03μs     0.17  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'h')
-     6.84±0.02μs      1.05±0.01μs     0.15  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'B')
-     6.83±0.01μs      1.05±0.01μs     0.15  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'B')
-     6.84±0.02μs      1.04±0.01μs     0.15  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'b')
-     6.83±0.01μs      1.03±0.01μs     0.15  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'b')
-        6.34±0μs          769±7ns     0.12  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'h')
-     6.33±0.01μs          766±4ns     0.12  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'H')
-        6.40±0μs          600±6ns     0.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'B')
-        6.40±0μs          598±5ns     0.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'b')

```
</details>


#### Binary size(striped)

| LIB         | Before(KBytes) | After(KBytes) | Diff(KBytes) |
| ----------- | ------------- | ------------ | ---------- |
| _multiarray_umath.cpython-38-x86_64-linux-gnu.so | 6124  | 6296 | 172 |

seiko2plus committed:
MAINT, SIMD: Removes compiler definitions of attribute-based CPU dispatching

seiko2plus committed:
ENH, SIMD: move auto-vectorized inner functions to new dispatchable source

seiko2plus committed:
MAINT, SIMD: fix c++ build when SSE intrinsics are in the scope

seiko2plus committed:
ENH, SIMD: dispatch the rest of all ufuncs with fast calls

seiko2plus committed:
MAINT, SIMD: fix c++ build when AVX2 intrinsics are in the scope

seiko2plus committed:
BLD, MESON: __builtin_prefetch is not limited to x86

seiko2plus committed:
BENCH: Clean up ufunc_strides

mattip commented on numpy/core/src/umath/loops_autovec.dispatch.c.src:
```diff
@@ -0,0 +1,248 @@
+NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(@TYPE@_positive)
+(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
+{
+    UNARY_LOOP_FAST(@type@, @type@, *out = +in);
```
Any idea why coverage marks these as not hit?

seiko2plus replied:
Not all data types are covered by test_umath.py, just basic contiguous tests. I'm working on implementing more deep testing units for ufunc/umath in a new path `test_umath_deep.py` since the current `test_umath`.py is too large.

mattip commented:
Looks like a nice code cleanup. Do benchmarks confirm that this does not change anything?

seiko2plus committed:
ENH, SIMD: reduce binary size of loops_autovec

seiko2plus committed:
BENCH: cover auto-vectorized ufunc operations

seiko2plus committed:
BENCH: Add new data generator

  for fairness and to stabilize the benchmark

seiko2plus committed:
SIMD: Disable auto-vectorization of avx512_skx

 Abandon the idea of providing binary objects for AVX512_SKX due to the massive increase
 in binary size over +400k (striped) with no vast change in performance except
 for reciprocal on some data types, tested against gcc/12.2.1 build:

      AVX512_SKX         AVX2                    ratio
     [e0e4acb7]       [0a56560e]
     <removes_old_cpu_dispatcher~1>       <removes_old_cpu_dispatcher>
+     26.0±0.07μs          125±7μs     4.83  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'L')
+     26.0±0.04μs          122±4μs     4.71  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'Q')
+     25.8±0.01μs        110±0.2μs     4.24  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'q')
+     25.9±0.08μs        108±0.6μs     4.18  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'l')
+     26.3±0.04μs      43.7±0.04μs     1.66  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'reciprocal'>, 1, 1, 'I')
+     1.63±0.03μs      2.39±0.02μs     1.47  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'l')
+     1.64±0.05μs         2.39±0μs     1.46  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'Q')
+     1.63±0.05μs      2.37±0.01μs     1.45  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'L')
+     1.64±0.04μs      2.37±0.01μs     1.45  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'q')
+     1.63±0.01μs      2.36±0.01μs     1.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'q')
+     1.63±0.01μs         2.35±0μs     1.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'L')
+     1.64±0.02μs      2.36±0.02μs     1.44  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'Q')
+     1.64±0.01μs         2.35±0μs     1.43  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'l')
+     8.69±0.07μs      12.5±0.03μs     1.43  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'l')
+     8.75±0.09μs      12.4±0.09μs     1.42  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'L')
+     8.65±0.09μs       12.3±0.1μs     1.42  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'Q')
+     8.63±0.02μs       12.2±0.2μs     1.42  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'q')
+        1.63±0μs      2.28±0.01μs     1.39  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
+        1.63±0μs         2.27±0μs     1.39  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
+        1.63±0μs         2.26±0μs     1.39  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'l')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'Q')
+     1.64±0.01μs         2.27±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'L')
+     1.72±0.06μs      2.37±0.01μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'q')
+     1.64±0.01μs         2.26±0μs     1.38  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
+     1.74±0.05μs         2.39±0μs     1.37  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
+     1.75±0.08μs      2.37±0.01μs     1.36  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
+     1.74±0.03μs      2.35±0.01μs     1.35  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'Q')
+     1.75±0.02μs      2.35±0.01μs     1.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'L')
+     1.75±0.04μs         2.36±0μs     1.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
+     1.77±0.05μs      2.37±0.01μs     1.34  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'l')
+     1.77±0.02μs         2.35±0μs     1.33  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'q')
+     1.69±0.01μs      2.20±0.01μs     1.31  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'l')
+     1.69±0.02μs      2.20±0.01μs     1.30  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'L')
+     1.69±0.02μs      2.21±0.01μs     1.30  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'Q')
+     1.70±0.02μs      2.21±0.01μs     1.30  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'q')
+     1.70±0.01μs      2.18±0.01μs     1.28  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'l')
+        1.69±0μs      2.15±0.01μs     1.27  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'L')
+     1.70±0.01μs      2.16±0.01μs     1.27  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'Q')
+     1.71±0.01μs      2.17±0.01μs     1.27  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'q')
+         886±7ns         1.10±0μs     1.24  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'H')
+         895±6ns         1.10±0μs     1.23  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'H')
+         897±7ns         1.10±0μs     1.23  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'h')
+     4.25±0.02μs      5.03±0.01μs     1.18  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'I')
+     1.31±0.02μs         1.54±0μs     1.18  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'I')
+     1.31±0.01μs         1.54±0μs     1.17  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'i')
+     4.27±0.08μs         4.95±0μs     1.16  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'i')
+        996±20ns         1.15±0μs     1.16  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'h')
+        1.21±0μs         1.40±0μs     1.15  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'b')
+     1.33±0.03μs      1.54±0.01μs     1.15  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
+        1.19±0μs         1.37±0μs     1.15  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'B')
+     1.34±0.01μs         1.53±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
+     1.32±0.02μs         1.51±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'I')
+        1.19±0μs         1.36±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'left_shift'>, 1, 1, 1, 'b')
+     1.32±0.01μs         1.50±0μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_and'>, 1, 1, 1, 'i')
+     1.36±0.01μs      1.54±0.01μs     1.14  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'h')
+        1.20±0μs         1.36±0μs     1.13  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'right_shift'>, 1, 1, 1, 'B')
+        1.43±0μs      1.61±0.01μs     1.13  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'H')
+        1.43±0μs         1.61±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'H')
+        1.38±0μs         1.55±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'h')
+     1.34±0.01μs         1.51±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
+     1.35±0.02μs         1.51±0μs     1.12  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
+      15.3±0.5μs      17.1±0.03μs     1.11  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'l')
+      15.3±0.4μs      17.1±0.06μs     1.11  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'q')
+        1.05±0μs         1.16±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'i')
+     1.41±0.01μs         1.57±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'b')
+     1.34±0.01μs         1.49±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'I')
+        1.34±0μs      1.48±0.01μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'i')
+     1.05±0.01μs         1.16±0μs     1.11  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_and'>, 1, 1, 1, 'I')
+     1.05±0.01μs         1.16±0μs     1.10  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'I')
+     1.43±0.02μs      1.58±0.01μs     1.10  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'b')
+      15.0±0.3μs      16.6±0.05μs     1.10  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'Q')
+      15.1±0.2μs       16.5±0.1μs     1.10  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'l')
+      15.0±0.3μs      16.5±0.02μs     1.10  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'l')
+        1.06±0μs         1.16±0μs     1.10  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_xor'>, 1, 1, 1, 'i')
+     1.50±0.01μs         1.64±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'right_shift'>, 1, 1, 1, 'B')
+     1.51±0.01μs         1.64±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'left_shift'>, 1, 1, 1, 'B')
+        1.60±0μs      1.74±0.01μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'L')
+      15.1±0.5μs      16.4±0.03μs     1.09  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'q')
+     1.34±0.01μs         1.45±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'I')
+        1.60±0μs         1.74±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'Q')
+        1.60±0μs         1.74±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'q')
+        1.61±0μs         1.74±0μs     1.09  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'logical_or'>, 1, 1, 1, 'l')
+      7.23±0.3μs      7.84±0.06μs     1.08  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'sign'>, 1, 1, 'i')
+      15.1±0.3μs      16.4±0.05μs     1.08  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'L')
+     3.28±0.04μs      3.54±0.01μs     1.08  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'h')
+      15.1±0.2μs      16.2±0.02μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'invert'>, 1, 1, 'q')
+     1.14±0.01μs      1.23±0.01μs     1.07  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'h')
+     1.35±0.01μs      1.44±0.01μs     1.07  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'logical_or'>, 1, 1, 1, 'i')
+      15.1±0.1μs      16.2±0.04μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'q')
+      7.09±0.1μs       7.59±0.1μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'positive'>, 1, 1, 'i')
+     2.60±0.03μs      2.78±0.04μs     1.07  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'h')
+     3.29±0.07μs      3.50±0.07μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'conjugate'>, 1, 1, 'H')
+        1.16±0μs      1.23±0.01μs     1.06  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_and'>, 1, 1, 1, 'H')
+     2.57±0.02μs      2.73±0.01μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'logical_not'>, 1, 1, 'H')
+     3.29±0.07μs      3.49±0.03μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'absolute'>, 1, 1, 'H')
+        1.18±0μs      1.24±0.01μs     1.06  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'H')
+     6.94±0.04μs       7.32±0.1μs     1.06  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'conjugate'>, 1, 1, 'I')
+     1.16±0.01μs         1.23±0μs     1.05  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_xor'>, 1, 1, 1, 'h')
+     3.17±0.03μs      3.33±0.06μs     1.05  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'positive'>, 1, 1, 'h')
+        1.15±0μs      1.21±0.01μs     1.05  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'h')
+     1.15±0.01μs         1.21±0μs     1.05  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in0(<ufunc 'logical_or'>, 1, 1, 1, 'H')
-     1.04±0.02μs          992±3ns     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'bitwise_and'>, 1, 1, 1, 'b')
-     1.00±0.01μs          950±1ns     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'b')
-     2.34±0.03μs      2.22±0.01μs     0.95  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'bitwise_xor'>, 1, 1, 1, 'l')
-     2.09±0.05μs         1.98±0μs     0.95  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'B')
-      7.29±0.1μs      6.90±0.01μs     0.95  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'i')
-     2.10±0.06μs         1.98±0μs     0.94  bench_ufunc_strides.UnaryIntContig.time_unary(<ufunc 'square'>, 1, 1, 'b')
-     2.34±0.03μs      2.20±0.01μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'right_shift'>, 1, 1, 1, 'l')
-     2.24±0.02μs      2.10±0.01μs     0.94  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'subtract'>, 1, 1, 1, 'Q')
-         639±2ns        600±0.9ns     0.94  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'multiply'>, 1, 1, 1, 'b')
-         640±7ns        596±0.6ns     0.93  bench_ufunc_strides.BinaryIntContig.time_binary(<ufunc 'multiply'>, 1, 1, 1, 'B')
-     2.27±0.08μs         2.10±0μs     0.93  bench_ufunc_strides.BinaryIntContig.time_binary_scalar_in1(<ufunc 'subtract'>, 1, 1, 1, 'l')

seiko2plus committed:
MAINT, SIMD: fix c++ build when VSX intrinsics are in the scope

seiko2plus committed:
SIMD: Suppress VSX ambiguous warnings

seiko2plus commented:
> Looks like a nice code cleanup. Do benchmarks confirm that this does not change anything?

I had to enhance the benchmark code to answer this question. No regression, please check the pr description, I have updated it.

charris commented:
Thanks Sayed. I figure the earlier this big change gets in, the better.
